test/unit/src/renderers/shaders/UniformsUtils.tests.js

```diff
@@ -29,12 +38,114 @@ export default QUnit.module( 'Renderers', () => {
+					assert.ok( uniforms.arrayValue.value[i] === uniformClones.arrayValue.value[i] );
+				}
+			} );
 
-				assert.ok( false, 'everything\'s gonna be alright' );
+			QUnit.test( 'cloneUniforms clones properties', ( assert ) => {
+
+				const uniforms = {
+					floatValue: { value: 1.23 },
+					intValue: { value: 1 },
+					boolValue: { value: true },
+					colorValue: { value: new Color(0xFF00FF) },
+					vector2Value: { value: new Vector2(1, 2) },
+					vector3Value: { value: new Vector3(1, 2, 3) },
+					vector4Value: { value: new Vector4(1, 2, 3, 4) },
+					matrix3Value: { value: new Matrix3() },
+					matrix4Value: { value: new Matrix4() },
+					quatValue: { value: new Quaternion(1, 2, 3, 4) },
+					arrayValue: { value: [1, 2, 3, 4] },
+					textureValue: { value: new Texture(null, CubeReflectionMapping) },
+				};
+	
+				const uniformClones = UniformsUtils.clone(uniforms);
+
+				// Modify the originals
+				uniforms.floatValue.value = 123.0;
+				uniforms.intValue.value = 123;
+				uniforms.boolValue.value = false;
+				uniforms.colorValue.value.r = 123.0;
+				uniforms.vector2Value.value.x = 123.0;
+				uniforms.vector3Value.value.x = 123.0;
+				uniforms.vector4Value.value.x = 123.0;
+				uniforms.matrix3Value.value.elements[0] = 123.0;
+				uniforms.matrix4Value.value.elements[0] = 123.0;
+				uniforms.quatValue.value.x = 123.0;
+				uniforms.arrayValue.value[0] = 123.0;
+				uniforms.textureValue.value.mapping = UVMapping;
+
+				assert.ok( uniforms.floatValue.value !== uniformClones.floatValue.value );
+				assert.ok( uniforms.intValue.value !== uniformClones.intValue.value );
+				assert.ok( uniforms.boolValue.value !== uniformClones.boolValue.value );
+				assert.ok( !uniforms.colorValue.value.equals(uniformClones.colorValue.value) );
+				assert.ok( !uniforms.vector2Value.value.equals(uniformClones.vector2Value.value) );
+				assert.ok( !uniforms.vector3Value.value.equals(uniformClones.vector3Value.value) );
+				assert.ok( !uniforms.vector4Value.value.equals(uniformClones.vector4Value.value) );
+				assert.ok( !uniforms.matrix3Value.value.equals(uniformClones.matrix3Value.value) );
+				assert.ok( !uniforms.matrix4Value.value.equals(uniformClones.matrix4Value.value) );
+				assert.ok( !uniforms.quatValue.value.equals(uniformClones.quatValue.value) );
+				assert.ok( uniforms.textureValue.value.mapping !== uniformClones.textureValue.value.mapping );
+				assert.ok( uniforms.arrayValue.value[0] !== uniformClones.arrayValue.value[0] );
+
+				// Texture source remains same
+				assert.ok( uniforms.textureValue.value.source.uuid === uniformClones.textureValue.value.source.uuid );
 
 			} );
 
+			QUnit.test( 'cloneUniforms skips render target textures', ( assert ) => {
+
+				const uniforms = {
+					textureValue: { value: new Texture(null, CubeReflectionMapping) },
+				};
+
+				uniforms.textureValue.value.isRenderTargetTexture = true;
+
+				let warningLogged = false;
```

Comment:
@Mugen87 sorry, im wrong.  Yes, that will cause issues.  We have something that turns off warnings.  This is not it.

Reply:
In this case, I think it's okay to remove the entire code block. It's not necessary for now to test whether the warning has been logged or not.

Reply:
@simondevyoutube 

You've been using the structure for keeping the console clean when a test will make it messy.

```js
			// surpress the following console message during testing
			// THREE.DataUtils.toHalfFloat(): Value out of range.

			console.level = CONSOLE_LEVEL.OFF;
			assert.ok( DataUtils.toHalfFloat( 100000 ) === 31743, 'Passed!' );
			assert.ok( DataUtils.toHalfFloat( - 100000 ) === 64511, 'Passed!' );
			console.level = CONSOLE_LEVEL.DEFAULT;
```

There's a hooks parameter, where you can do a "beforeeach and aftereach" if something like this is critical.  Search hooks in the tests to see examples.  Use these if you want to keep this approach and use them to clean up.

https://api.qunitjs.com/QUnit/hooks/

The other option, like mugen said, we are not going hard on testing warnings.  Once this becomes mature, we would be looking at tooling that handles this and other bits gracefully.

Reply:
Too quick again, you don't want those hooks referenced above.  You want the module level hooks.

Add a quick `hooks` parameter to the arrow function to get module level hooks.

```js
		QUnit.module( 'CubicBezierCurve', ( hooks ) => {

			let curve = undefined;
			hooks.before( function () {

				curve = new CubicBezierCurve(
					new Vector2( - 10, 0 ),
					new Vector2( - 5, 15 ),
					new Vector2( 20, 15 ),
					new Vector2( 10, 0 )
				);

			} );
```

Reply:
Updated the code to the `DataUtils` appraoch.